ytypes/string_type.go

```diff
@@ -26,6 +27,57 @@ import (
 
 // Refer to: https://tools.ietf.org/html/rfc6020#section-9.4.
 
+var (
+	// posixRegexCache stores previously-compiled Regexp objects.
+	// This helps the performance of validation of, say, a large prefix
+	// list that have the same pattern specification.
+	//
+	// Concurrency Requirements
+	//
+	// Only the regexp cache has to be protected by mutexes, since
+	// a Regexp is safe for concurrent use by multiple goroutines:
+	// https://golang.org/src/regexp/regexp.go
+	posixRegexCache = map[string]*regexp.Regexp{}
+	posixRegexMutex = &sync.RWMutex{}
+	re2RegexCache   = map[string]*regexp.Regexp{}
+	re2RegexMutex   = &sync.RWMutex{}
+)
+
+// compilePattern returns the compiled regex for the given regex
+// pattern. It caches previous look-ups for faster performance.
+// Go's regexp implementation might be relatively slow compared to other
+// languages: https://github.com/golang/go/issues/11646
+func compilePattern(pattern string, isPOSIX bool) (*regexp.Regexp, error) {
+	regexCache := re2RegexCache
+	regexMutex := re2RegexMutex
+	regexCompile := regexp.Compile
+	if isPOSIX {
+		regexCache = posixRegexCache
+		regexCompile = regexp.CompilePOSIX
+		regexMutex = posixRegexMutex
+	}
+
+	regexMutex.RLock()
+	r, ok := regexCache[pattern]
+	regexMutex.RUnlock()
```

Comment:
Prefer to use `defer` here so that we don't end up changing this code in some way that gets us in lock contention. Here I think the usual style is to do this through an inline-function:

```golang
r := func() *regexp.Regexp {
   regexpMutex.RLock()
   defer regexpMutex.RUnlock()
   return regexpCache[pattern]
}()

if r != nil {
   return r, nil
}

re, err := regexpCompile(pattern)
if err != nil {
 return nil, err
}
regexpMutex.Lock()
defer regexpMutex.Unlock()
regexpCache[pattern] = re

return re, nil 
```

Reply:
Done.